### PR TITLE
Initial support for external memory in gradient index.

### PR DESCRIPTION
**Pull request by trivialfis**
* Add hessian to batch param in preparasion of new approx impl.
* Extract a push method for gradient index matrix.
* Use span instead of vector ref for hessian in sketching.
* Create a binary format for gradient index.

### CODE REVIEW

**Comment by trivialfis** [comment]
@ShvetsKS Would you like to take a look into this PR?

**Review by RAMitchell — COMMENTED** [review]
Looks fine to me but will need approval from @ShvetsKS.

**Review by ShvetsKS — COMMENTED** [review]
Sorry for long response.
LGTM in general, only one small comment.

File: src/tree/updater_quantile_hist.cc [hunk 1]
```diff
@@ -69,13 +69,13 @@ void QuantileHistMaker::CallBuilderUpdate(const std::unique_ptr<Builder<Gradient
...
                                DMatrix *dmat,
                                const std::vector<RegTree *> &trees) {
-  auto const &gmat =
-      *(dmat->GetBatches<GHistIndexMatrix>(
-                BatchParam{GenericParameter::kCpuId, param_.max_bin})
-            .begin());
+  auto it = dmat->GetBatches<GHistIndexMatrix>(
+                    BatchParam{GenericParameter::kCpuId, param_.max_bin})
+                .begin();
+  auto p_gmat = it.Page();
```
[the remark this answers is not in the document]
Comment (trivialfis) [reply src/tree/updater_quantile_hist.cc:75]: @ShvetsKS The PR only adds the support in `GHistIndexMatrix`.  None of the existing tree methods can make use of it yet.  This PR creates a new approx tree method that uses it https://github.com/dmlc/xgboost/pull/7214 .  After that, we can make hist support external memory properly.  Right now for hist, the gradient index is concatenated into a single page.

Also, see https://github.com/dmlc/xgboost/pull/7214#issuecomment-917618392 .

**Pull request merged**